### PR TITLE
Fix incorrect "Saving latest checkpoint" warning

### LINKED ISSUE #3578: Incorrect "Saving latest checkpoint" warning
## 🐛 Bug

"Saving latest checkpoint..." warning appears regardless of whether a `ModelCheckpoint` exists or `save_last` is set to `True`

https://github.com/PyTorchLightning/pytorch-lightning/blob/a71d62d8409f4960a4b438b8d19c924d3636c73f/pytorch_lightning/trainer/training_loop.py#L167-L169

https://github.com/PyTorchLightning/pytorch-lightning/blob/a71d62d8409f4960a4b438b8d19c924d3636c73f/pytorch_lightning/trainer/training_loop.py#L196-L204

This might confuse an user to think the last checkpoint got saved when it did not.

## Proposed change:

```python
def check_checkpoint_callback(self, should_check_val, force_save=False):
    model = self.trainer.get_model()

    # when no val loop is present or fast-dev-run still need to call checkpoints
    # TODO bake this logic into the checkpoint callback
    should_activate = not is_overridden('validation_step', model) and not should_check_val
    if should_activate or force_save:
        checkpoint_callbacks = [c for c in self.trainer.callbacks if isinstance(c, ModelCheckpoint)]
        if any(c.save_last for c in checkpoint_callbacks):
            rank_zero_warn('Saving latest checkpoint..')
        [c.on_validation_end(self.trainer, model) for c in checkpoint_callbacks]
```

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

- Only shows warning when appropriate
- Adds a test
- Cleans up imports

Fixes #3578

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

### CODE REVIEW

File: tests/callbacks/test_model_checkpoint.py [hunk 1]
```diff
@@ -243,3 +243,24 @@ def test_ckpt_metric_names_results(tmpdir):
...
+        trainer.fit(model)
+    # make sure warning does not appear if save_last=False:
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        checkpoint_callback=ModelCheckpoint(filepath=tmpdir, save_top_k=0, save_last=False),
+        max_epochs=1,
+    )
+    with pytest.warns(None) as record:
+        trainer.fit(model)
+    assert not any(r.message.args[0] == "Saving latest checkpoint..." for r in record)
```
Comment (Borda) [inline tests/callbacks/test_model_checkpoint.py:266]:
```suggestion
    assert not any("Saving latest checkpoint..." in r.message.args[0] for r in record)
```

File: pytorch_lightning/trainer/training_loop.py [hunk 2]
```diff
@@ -201,6 +200,8 @@ def check_checkpoint_callback(self, should_check_val, force_save=False):
         should_activate = not is_overridden('validation_step', model) and not should_check_val
         if should_activate or force_save:
             checkpoint_callbacks = [c for c in self.trainer.callbacks if isinstance(c, ModelCheckpoint)]
+            if any(c.save_last for c in checkpoint_callbacks):
+                rank_zero_warn('Saving latest checkpoint...')
```
Comment (Borda) [inline pytorch_lightning/trainer/training_loop.py:204]: just not sure what warn and just info message...

Comment (carmocca) [reply pytorch_lightning/trainer/training_loop.py:204]: It was a warning previously, so I kept it that way. Tell me if you want me to change it

File: tests/callbacks/test_model_checkpoint.py (hunk 1 above)
Comment (carmocca) [reply tests/callbacks/test_model_checkpoint.py:266]: Did you mean?
```python
assert not any("Saving latest checkpoint..." in r.message.args for r in record)
```

File: pytorch_lightning/trainer/training_loop.py (hunk 2 above)
Comment (Borda) [reply pytorch_lightning/trainer/training_loop.py:204]: I see, I would change it... @awaelchli?

Comment (awaelchli) [reply pytorch_lightning/trainer/training_loop.py:204]: I vote for change to info, but no strong feeling, either way is fine for me

Comment (carmocca) [reply pytorch_lightning/trainer/training_loop.py]: Done

**Comment by awaelchli** [comment]
I think we should clarify whether this "forced" action of saving the checkpoint is meant to override the ModelCheckpoint argument "save_last" or not. I'm not sure, so we need to ask @williamFalcon

**Comment by carmocca** [comment]
> whether this "forced" action of saving the checkpoint is meant to override the ModelCheckpoint argument "save_last" or not.

Why would it? This PR does not attempt to force `save_last` when training is interrupted. It just fixes the warning.

There is no forced action. `check_checkpoint_callback` just runs `on_validation_end` to give `ModelCheckpoint` a chance to save.

**Comment by awaelchli** [comment]
Yes I know. But you see, when you moved the warning from "on train end" into the method for checking the modelcheckpoint you have now changed the meaning of this message. While on_train_end is only called when training ends or gets stopped with Ctrl+C, the check_checkpoint_callback is done at every epoch. The forced action I was referring to is the argument `force_save=True`, which does not do what it is supposed to.

**Comment by carmocca** [comment]
@awaelchli's comment made me realize that what I did was not correct. With the latest commit, the meaning of the message is kept.

It refactors `check_checkpoint_callback` and removes an unused class variable `self.should_check_val` which was not getting set anywhere.

File: pytorch_lightning/trainer/training_loop.py [hunk 3]
```diff
@@ -502,8 +498,8 @@ def run_training_epoch(self):
             self.num_optimizers
         )
 
-        # checkpoint callback
-        self.check_checkpoint_callback(self.should_check_val)
+        # when no val loop is present or fast-dev-run still need to call checkpoints
+        self.check_checkpoint_callback(not (should_check_val and is_overridden('validation_step', model)))
```
Comment (awaelchli) [inline pytorch_lightning/trainer/training_loop.py:502]: just for my understanding, this is here in case in the above loop we never run the validation loop (because no val step or not right period). In this case, we still want to run it for the training metric, in case it is monitoring training. You are adding this "not" clause because you also don't want to run it twice. 
Is my reasoning correct?

**Review by awaelchli — APPROVED** [review]
Looks good. Pulled the branch and ran a few scripts to see how checkpoints are saved, and it seems to work as intended. 
Just need some clarification/confirmation about the comment above. 
cheers

File: pytorch_lightning/trainer/training_loop.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/trainer/training_loop.py:502]: The comment and the condition (`not (should_check_val and is_overridden('validation_step', model))`) were inside of the `check_checkpoint_callback` function, I moved it out because I think it is cleaner to check it outside of the function.

I think your reasoning is correct but @williamFalcon is the one who wrote it originally. Maybe he can confirm it.

Comment (awaelchli) [reply pytorch_lightning/trainer/training_loop.py:502]: ok, just be aware that above we have `not A and not B`, which is equivalent to `not (A or B)` but here we have `not (A and B)` which is not the same.

Comment (carmocca) [reply pytorch_lightning/trainer/training_loop.py:502]: Good catch! Messed up de morgan's law lol. I'll fix it

**Review by Borda — APPROVED** [review]
LGTM

**Review by SkafteNicki — APPROVED** [review]
LGTM
